Add translation support

zSoulweaver opened the pull request:
Adds i18n support with help of nuxt-i18n module.

Added support for blog posts by having a `locales` folder within the main blog folder. This keeps everything for that particular blog post in one directory. If the current locale does not have a translation available it will fallback to English. The locales folder contains the same content as the `index.md` file although the file name is changed to the language code it is being translated for.

Language can be switched in the navbar, it will automatically populate the list of available locales. The locales will be defined within the `nuxt.config.js` file and translations will be located in the `locale` folder.

zSoulweaver committed:
fix: change navbar to be cross browser compat
- Removes arrow/triangle
- Centers dropdown with nav item

zSoulweaver committed:
feat: add i18n module
- adds two example files
- Spanish file is not translated, just for testing

zSoulweaver committed:
feat: add language switcher to navbar

zSoulweaver committed:
feat: convert index to using translations
- moves nuxt-links to proper format for use with i18n module

zSoulweaver committed:
feat: move tool descriptions to locale files

zSoulweaver committed:
feat: blog posts can be written in other langs
- falls back to english article if no translation available

zSoulweaver committed:
refactor: change lang folder to locale

zSoulweaver committed:
Merge branch 'master' into i18n

zSoulweaver committed:
feat: moved more pages to i18n
- Removed page names out of yaml files into locale files instead

zSoulweaver committed:
feat: add i18n support for projects
- same implementation as blog posts since description is markdown

zSoulweaver committed:
feat: add i18n support for page footer

zSoulweaver committed:
feat: add published text to locales

zSoulweaver commented:
*Hopefully* added most things that can be translated to the locales file.

Some changes have moved the page titles and descriptions out of the yaml files in the content directory to the locales directory. No need to double handle things. For example in the project.yaml file we had the following data:
```yaml
title: Projects
subtitle: All the projects by the team and community
```
This is now sitting in the locale file under
```js
projects: {
  pageTitle: 'Projects',
  pageDescription: 'All the projects by the team and community'
}
```

This is a reoccurring change throughout all the files that had the page titles, descriptions, paragraphs etc defined within the yaml files.

Also added support for project files to be localised, as these were markdown files it's not as easy to just fine the content in the locale file with everything else. Similar to how the blog posts were handled it now looks for a `locale` folder and then looks for the current language code, if it can't find that code it simply defaults back to the English file. Along with this change the English project files in the root folder **MUST** be named `index.md` or they will not be picked up.

The footer is also now localised, in the yaml it utilises either a `name` or `nameLocale` key. The name key will just display exactly whatever string of text is defined whereas nameLocale will do a lookup in the locale file. For example:
```js
footer: {
    main: {
      label: 'Main',
      home: 'Home'
    },
  }
```
```yaml
columns:
  - title: footer.main.label # => Main
    items:
      - nameLocale: footer.main.home # => Home
        link: '/'
      - nameLocale: nav.blog # uses same translations from nav as they should be the same => Blog
        link: '/blog'
      - name: Projects # doesn't reference the locale file, will render that string => Projects
        link: '/projects'
```

If there are any other parts of the website that should be translated let me know, I'm pretty sure I've covered most of them.

zSoulweaver commented:
If everything looks good, before merge I'll remove the dummy translations I did for testing.

Akiway commented on pages/blog/_slug.vue:
```diff
@@ -17,14 +17,30 @@
+    let [post] = await $content('blog', { deep: true })
+      .where({
+        dir: isEnglish ? `/blog/${params.slug}` : `/blog/${params.slug}/locales`,
+        slug: isEnglish ? 'index' : currentLocale
```
Put the english file as any other lang file to keep it consistent.
```
-BlogName
    -image.jpg
    -locales
        -en.md <=
        -es.md
        -fr.md
```

So you endup with a code like this:
```suggestion
        dir: `/blog/${params.slug}/locales`,
        slug: currentLocale
```

zSoulweaver replied:
Done.

Akiway commented on nuxt.config.js:
```diff
@@ -64,9 +64,31 @@ export default {
+        name: 'Espanol'
+      }
+    ],
+    defaultLocale: 'en',
+    vueI18n: {
+      fallbackLocale: 'en'
```
use a environnement variable like `DEFAULT_LANG`
```suggestion
    defaultLocale: process.env.DEFAULT_LANG,
    vueI18n: {
      fallbackLocale: process.env.DEFAULT_LANG
```

In the same idea, you can also replace `'en'` line 11 with `process.env.DEFAULT_LANG`

zSoulweaver replied:
Done, used `DEFAULT_LOCALE` to keep it consistent.

Akiway commented on pages/blog/_slug.vue:
```diff
@@ -17,14 +17,30 @@
-      .where({ dir: `/blog/${params.slug}` })
+  async asyncData ({ $content, params, error, app }) {
+    const currentLocale = app.i18n.locale
+    const isEnglish = currentLocale === 'en'
```
Use environment variable and maybe even rename the variable:
```suggestion
    const isDefault = currentLocale === process.env.DEFAULT_LANG
```

zSoulweaver replied:
Done.

Akiway commented on pages/blog/index.vue:
```diff
@@ -33,12 +34,51 @@ export default {
         error({ statusCode: 404, message: 'Page not found' })
       })
 
+    if (currentLocale !== 'en') {
```
```suggestion
    if (currentLocale !== process.env.DEFAULT_LANG) {
```

zSoulweaver replied:
Done.

Akiway commented on pages/projects/index.vue:
```diff
@@ -46,14 +48,77 @@ export default {
       .fetch()
       .catch(() => {})
 
+    if (currentLocale !== 'en') {
```
```suggestion
    if (currentLocale !== process.env.DEFAULT_LANG) {
```

zSoulweaver replied:
Done.

Akiway commented on pages/blog/index.vue:
```diff
@@ -33,12 +34,51 @@ export default {
       .fetch()
       .catch(() => {
         error({ statusCode: 404, message: 'Page not found' })
       })
```
An idea:
Always try to load the currentLang file, and load the defaultLang file in the catch ? So it prevent from loading unecessary english files that are translated.

zSoulweaver replied:
In terms of performance this would only really be affecting the build/generation time of the site and shouldn't affect the performance in production. Although I haven't tested that.

Will make the change anyway as it'll probably look cleaner overall.

zSoulweaver replied:
Didn't test, but with how the content module handles this particular case it won't actually catch an error, it just returns an empty array IIRC.

Akiway reviewed (changes requested):
Create a .env file so you can add `DEFAULT_LANG=en`
and use it everywhere like `=== process.env.DEFAULT_LANG` instead of doing `=== 'en'`

kpracuk reviewed (approved):
As suggested earlier: `.env` would be pretty good in this case, but other than that lgtm.

zSoulweaver committed:
refactor: change hardcoded default locale to env
- will fallback to 'en' if no env value is specified

squash

zSoulweaver committed:
refactor: fix typo

zSoulweaver commented:
Made above requested changes, used `DEFAULT_LOCALE` for environment variable naming to keep it consistent. Added env to GitHub workflow too, although in case an env is never specified it will fallback to `'en'`.

Moved the default English index.md file to the locales folder for consistency, this in turn broke some stuff in regards to image paths and such, and *should* be fixed.

zSoulweaver committed:
cleanup: remove Spanish test translations
